src/components/Table/GridTable.tsx

```diff
@@ -270,141 +259,73 @@ export function GridTable<R extends Kinded, X extends Only<GridTableXss, X> = an
+        visibleDataRows.push(tuple);
+      }
+    });
 
-    // Flatten the tuple tree into lists of rows
-    function visitRows(tuples: ParentChildrenTuple<R>[], level: number): void {
-      tuples.forEach(([row, children]) => {
-        if (row.kind === "header") {
-          headerRows.push([row, makeRowComponent(row, level)]);
-        } else if (row.kind === "totals") {
-          totalsRows.push([row, makeRowComponent(row, level)]);
-        } else if (row.kind === "expandableHeader") {
-          expandableHeaderRows.push([row, makeRowComponent(row, level)]);
-        } else {
-          visibleDataRows.push([row, makeRowComponent(row, level)]);
-          // tuples has already been client-side filtered, so just check collapsed
-          if (children.length && !collapsedIds.includes(row.id)) {
-            visitRows(children, level + 1);
-          }
-        }
-      });
-    }
+    // Once our header rows are created we can organize them in expected order.
+    const tableHeadRows = expandableHeaderRows.concat(headerRows).concat(totalsRows);
 
-    // Call `visitRows` with our post-filtered list
-    const [filteredRowIds, filteredRows] = filterRows(api, columns, maybeSorted, filter);
-    visitRows(filteredRows, 0);
-
-    // Check for any selected rows that are not displayed in the table because they don't
-    // match the current filter, or are no longer part of the `rows` prop. We persist these
-    // selected rows and hoist them to the top of the table.
-    if (sortedKeptSelections.length) {
-      keptSelectedRows.push([keptGroupRow as GridDataRow<R>, makeRowComponent(keptGroupRow as GridDataRow<R>, 1)]);
-      if (!collapsedIds.includes(KEPT_GROUP)) {
-        keptSelectedRows.push(
-          ...sortedKeptSelections.map((row, idx) => {
-            const isLast = idx === sortedKeptSelections.length - 1;
-            return [row, makeRowComponent(row, 1, true, isLast)] as RowTuple<R>;
-          }),
-        );
-      }
+    const tooManyClientSideRows = !!filterMaxRows && visibleDataRows.length > filterMaxRows;
+    if (tooManyClientSideRows) {
+      visibleDataRows = visibleDataRows.slice(0, filterMaxRows + keptSelectedRows.length);
     }
 
-    return [headerRows, visibleDataRows, totalsRows, expandableHeaderRows, keptSelectedRows, filteredRowIds];
-  }, [
-    as,
-    api,
-    filter,
-    maybeSorted,
-    columns,
-    style,
-    rowStyles,
-    maybeStyle,
-    columnSizes,
-    collapsedIds,
-    getCount,
-    keptGroupRow,
-    sortedKeptSelections,
-  ]);
-
-  // Once our header rows are created we can organize them in expected order.
-  const tableHeadRows = expandableHeaderRows.concat(headerRows).concat(totalsRows);
-
-  const tooManyClientSideRows = filterMaxRows && visibleDataRows.length > filterMaxRows;
-  if (tooManyClientSideRows) {
-    visibleDataRows = visibleDataRows.slice(0, filterMaxRows + keptSelectedRows.length);
-  }
-
-  useEffect(() => {
-    tableState.setMatchedRows(filteredRowIds);
```

Comment:
We don't push matched-ness into the table state anymore, each `RowState.isMatched` does its own reactive calc, instead of just being a boolean that is set by some outside code.